halo/evmengine/keeper/abci_internal_test.go

```diff
@@ -0,0 +1,544 @@
+		require.NoError(t, err)
+		require.NotNil(t, resp)
+
+		tx, err := txConfig.TxDecoder()(resp.Txs[0])
+		require.NoError(t, err)
+
+		for _, msg := range tx.GetMsgs() {
+			if _, ok := msg.(*etypes.MsgExecutionPayload); ok {
+				assertExecutablePayload(t, msg, ts.Unix(), nextBlock.Hash(), ap, uint64(req.Height))
+			}
+		}
+	})
+
+	t.Run("build optimistic", func(t *testing.T) {
+		t.Parallel()
+		ctx, storeService := setupCtxStore(t)
+		cdc := getCodec(t)
+		txConfig := authtx.NewTxConfig(cdc, nil)
+
+		me, err := ethclient.NewEngineMock()
+		require.NoError(t, err)
+		mockEngine := MockEngineAPI{
+			Mock:   me,
+			fuzzer: ethclient.NewFuzzer(time.Now().Truncate(time.Hour * 24).Unix()),
+		}
+		ap := MockAddressProvider{}
+		keeper := NewKeeper(cdc, storeService, &mockEngine, txConfig, ap)
+
+		ts := time.Now()
+		latestHeight, err := mockEngine.BlockNumber(ctx)
+		require.NoError(t, err)
+		latestBlock, err := mockEngine.BlockByNumber(ctx, big.NewInt(int64(latestHeight)))
+		require.NoError(t, err)
+
+		mockEngine.pushPayload(t, ctx, ap.LocalAddress(), latestBlock.Hash(), ts)
+		nextBlock, blockPayload := mockEngine.nextBlock(t, latestHeight+1, uint64(ts.Unix()), latestBlock.Hash(), ap.LocalAddress())
+		_, err = mockEngine.Mock.NewPayloadV2(ctx, blockPayload)
+		require.NoError(t, err)
+		mockEngine.pushPayload(t, ctx, ap.LocalAddress(), nextBlock.Hash(), ts)
+
+		keeper.mutablePayload.UpdatedAt = time.Now()
+
+		req := &abci.RequestPrepareProposal{
+			Txs:    nil,
+			Height: int64(2),
+			Time:   time.Now(),
+		}
+
+		resp, err := keeper.PrepareProposal(ctx, req)
+		require.NoError(t, err)
+		require.NotNil(t, resp)
+
+		tx, err := txConfig.TxDecoder()(resp.Txs[0])
+		require.NoError(t, err)
+
+		for _, msg := range tx.GetMsgs() {
+			if _, ok := msg.(*etypes.MsgExecutionPayload); ok {
+				assertExecutablePayload(t, msg, req.Time.Unix()+1, nextBlock.Hash(), ap, uint64(req.Height))
+			}
+		}
+	})
+}
+
+func assertExecutablePayload(t *testing.T, msg sdk.Msg, ts int64, blockHash common.Hash, ap MockAddressProvider, height uint64) {
+	t.Helper()
+	executionPayload, ok := msg.(*etypes.MsgExecutionPayload)
+	require.True(t, ok)
+	require.NotNil(t, executionPayload)
+	var ep *eengine.ExecutableData
+	err := json.Unmarshal(executionPayload.GetData(), &ep)
+	require.NoError(t, err)
+	require.Equal(t, int64(ep.Timestamp), ts)
+	require.Equal(t, ep.Random, blockHash)
+	require.Equal(t, ep.FeeRecipient, ap.LocalAddress())
+	require.Empty(t, ep.Withdrawals)
+	require.Equal(t, ep.Number, height)
+}
+
+func setupCtxStore(t *testing.T) (sdk.Context, store.KVStoreService) {
+	t.Helper()
+	key := storetypes.NewKVStoreKey("test")
+	storeService := runtime.NewKVStoreService(key)
+	testCtx := testutil.DefaultContextWithDB(t, key, storetypes.NewTransientStoreKey("transient_test"))
+	ctx := testCtx.Ctx.WithBlockHeader(cmtproto.Header{Time: cmttime.Now()})
+
+	return ctx, storeService
+}
+
+func getCodec(t *testing.T) *codec.ProtoCodec {
```

Comment:
What about rather:
```
moduletestutil "github.com/cosmos/cosmos-sdk/types/module/testutil"
...
moduletestutil.MakeTestEncodingConfig().Codec

```

Reply:
Doesn't help that much as we still have to register all the interfaces for the codec